src/auth/discord.rs

```diff
@@ -1 +1,404 @@
+    // API Docs: https://discord.com/developers/docs/getting-started
+    //API EndPoints: https://github.com/GregTCLTK/Discord-Api-Endpoints/blob/master/Endpoints.md
+    // TODO: It can be stored in the request cookies
+    pub async fn user_details(token: &str) -> fpm::Result<(String, String)> {
+        // API Docs: https://discord.com/api/users/@me
+        // TODO: Handle paginated response
+        #[derive(serde::Deserialize)]
+        struct UserDetails {
+            username: String,
+            id: String,
+        }
+        let user_obj: UserDetails = fpm::auth::utils::get_api(
+            "https://discord.com/api/users/@me",
+            format!("{}{}", "Bearer ", token).as_str(),
+        )
+        .await?;
+
+        Ok((user_obj.username, user_obj.id))
+    }
+
+    //This API will only be used to get access token for discord
+    pub async fn discord_token(url: &str, redirect_url: &str, code: &str) -> fpm::Result<String> {
+        let client_id = match std::env::var("DISCORD_CLIENT_ID") {
+            Ok(id) => id,
+            Err(_e) => {
+                println!("WARN: DISCORD_CLIENT_ID not set");
+                // TODO: Need to change this approach later
+                "FPM_TEMP_DISCORD_CLIENT_ID".to_string()
+            }
+        };
+        let client_secret = match std::env::var("DISCORD_CLIENT_SECRET") {
+            Ok(secret) => secret,
+            Err(_e) => {
+                println!("WARN: DISCORD_CLIENT_SECRET not set");
+                // TODO: Need to change this approach later
+                "FPM_TEMP_DISCORD_CLIENT_SECRET".to_string()
+            }
+        };
+        let mut map: std::collections::HashMap<&str, &str> = std::collections::HashMap::new();
+        map.insert("client_secret", client_secret.as_str());
+        map.insert("client_id", client_id.as_str());
+        map.insert("grant_type", "authorization_code");
+        map.insert("code", code);
+        map.insert("redirect_uri", redirect_url);
+
+        let response = reqwest::Client::new().post(url).form(&map).send().await?;
+
+        if !response.status().eq(&reqwest::StatusCode::OK) {
+            return Err(fpm::Error::APIResponseError(format!(
+                "DISCORD-API-ERROR: {}, Error: {}",
+                url,
+                response.text().await?
+            )));
+        }
+        let auth_obj = response.json::<DiscordAuthResp>().await?;
+        Ok(auth_obj.access_token)
+    }
+    pub async fn user_servers(token: &str) -> fpm::Result<Vec<String>> {
+        // API Docs: https://discord.com/api/users/@me/guilds
+        // TODO: Handle paginated response
+
+        #[derive(Debug, serde::Deserialize)]
+        struct UserGuilds {
+            name: String,
+        }
+        let user_server_list: Vec<UserGuilds> = fpm::auth::utils::get_api(
+            format!("{}?limit=100", "https://discord.com/api/users/@me/guilds").as_str(),
+            format!("{}{}", "Bearer ", token).as_str(),
+        )
```

Comment:
This API is paginated, please put // todo: we will implement it later.

Reply:
already mentioned